File: app/src/main/java/com/example/solar_alarm/Data/Converters.java
package com.example.solar_alarm.Data;

import android.os.Build;

import androidx.annotation.RequiresApi;
import androidx.room.TypeConverter;

import com.example.solar_alarm.Data.Enums.OffsetTypeEnum;
import com.example.solar_alarm.Data.Enums.SolarTimeTypeEnum;

import java.time.LocalDate;
import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;

@RequiresApi(api = Build.VERSION_CODES.O)
public class Converters
{
    private static DateTimeFormatter dateFormat = DateTimeFormatter.ISO_DATE;

    @TypeConverter
    public static LocalDate toLocalDate(String dateString) { return dateString == null ? null: LocalDate.parse(dateString); }

    @TypeConverter
    public static String fromLocalDate(LocalDate localDate) { return localDate == null ? null : localDate.format(dateFormat); }

    @TypeConverter
    public static int toOffsetTypeId(OffsetTypeEnum enumType) { return enumType.Id; }

    @TypeConverter
    public static OffsetTypeEnum toOffsetTypeEnum(int id) { return OffsetTypeEnum.values()[id]; }

    @TypeConverter
    public static int toSolarTimeTypeId(SolarTimeTypeEnum enumType) { return enumType.Id; }

    @TypeConverter
    public static SolarTimeTypeEnum toSolarTimeTypeEnum(int id) { return SolarTimeTypeEnum.values()[id]; }

    @TypeConverter
    public static String[] toTimeString(ZonedDateTime zonedDateTime)
    {
        // We will need to consider giving the user to choose their date and time formats.
        String hour;
        int hourInt = zonedDateTime.getHour() > 12 ? zonedDateTime.getHour() - 12 : zonedDateTime.getHour();

        if (hourInt < 10)
        {
            hour = String.format("%02d" , hourInt);
        }
        else
        {
            hour = String.valueOf(hourInt);
        }

        String ampm       = zonedDateTime.getHour() > 11 ? "PM" : "AM";
        String time       = hour + ":" + zonedDateTime.getMinute() + " " + ampm;

        //String dayOfWeek  = zonedDateTime.getDayOfWeek().toString().substring(0,3);
        String dayOfMonth = zonedDateTime.getDayOfMonth() < 10 ? String.format("%02d" , zonedDateTime.getDayOfMonth()) : String.valueOf(zonedDateTime.getDayOfMonth());
        String month      = zonedDateTime.getMonth().toString().substring(0,3);
        String date       = /*dayOfWeek + " " +*/ dayOfMonth + "-" + month + "-" + zonedDateTime.getYear();

        return new String[] {date, time};
    }
}

File: app/src/main/java/com/example/solar_alarm/Data/Converters.kt
package com.example.solar_alarm.data

import android.os.Build
import androidx.annotation.RequiresApi
import androidx.room.TypeConverter
import com.example.solar_alarm.data.enums.OffsetTypeEnum
import com.example.solar_alarm.data.enums.SolarTimeTypeEnum
import java.time.LocalDate
import java.time.OffsetDateTime
import java.time.ZonedDateTime
import java.time.format.DateTimeFormatter

@RequiresApi(api = Build.VERSION_CODES.O)
object Converters
{
    private val formatter = DateTimeFormatter.ISO_OFFSET_DATE_TIME

    @TypeConverter
    @JvmStatic
    fun toOffsetDateTime(value: String?): OffsetDateTime? {
        return value?.let {
            return formatter.parse(value, OffsetDateTime::from)
        }
    }

    @TypeConverter
    @JvmStatic
    fun fromOffsetDateTime(date: OffsetDateTime?): String? {
        return date?.format(formatter)
    }

    private val dateFormat = DateTimeFormatter.ISO_DATE
    @kotlin.jvm.JvmStatic
    @TypeConverter
    fun toLocalDate(dateString: String?): LocalDate? {
        return if (dateString == null) null else LocalDate.parse(dateString)
    }

    @kotlin.jvm.JvmStatic
    @TypeConverter
    fun fromLocalDate(localDate: LocalDate?): String? {
        return localDate?.format(dateFormat)
    }

    @kotlin.jvm.JvmStatic
    @TypeConverter
    fun toOffsetTypeId(enumType: OffsetTypeEnum): Int {
        return enumType.Id
    }

    @kotlin.jvm.JvmStatic
    @TypeConverter
    fun toOffsetTypeEnum(id: Int): OffsetTypeEnum {
        return OffsetTypeEnum.entries.find { it.Id == id } ?: OffsetTypeEnum.At
    }

    @kotlin.jvm.JvmStatic
    @TypeConverter
    fun toSolarTimeTypeId(enumType: SolarTimeTypeEnum): Int {
        return enumType.Id
    }

    @kotlin.jvm.JvmStatic
    @TypeConverter
    fun toSolarTimeTypeEnum(id: Int): SolarTimeTypeEnum {
        return SolarTimeTypeEnum.entries.find { it.Id == id } ?: SolarTimeTypeEnum.Sunrise
    }

    @TypeConverter
    fun toTimeString(zonedDateTime: ZonedDateTime): Array<String> {
        // We will need to consider giving the user to choose their date and time formats.
        val hour12 = when {
            zonedDateTime.hour == 0 -> 12
            zonedDateTime.hour > 12 -> zonedDateTime.hour - 12
            else -> zonedDateTime.hour
        }
        val ampm = if (zonedDateTime.hour >= 12) "PM" else "AM"
        val time = String.format(java.util.Locale.getDefault(), "%02d:%02d %s", hour12, zonedDateTime.minute, ampm)

        val dayOfMonth = String.format(java.util.Locale.getDefault(), "%02d", zonedDateTime.dayOfMonth)
        val month = zonedDateTime.month.toString().substring(0, 3)
        val date =  dayOfMonth + "-" + month + "-" + zonedDateTime.year
        return arrayOf(date, time)
    }
}